### ISSUE TITLE
Nier trips assert in aco_ssa_elimination.cpp

### ISSUE THREAD

**Issue by tarceri** [daniel-schuermann/mesa#24]
assert triggered in remove_merge_block(). Seems like the assert is there to capture more optimisation opportunities?

Dump of basic blocks triggering this code path:

```
BB11
/* logical preds: BB9, BB10, / linear preds: BB9, BB10,  */
	 s2: %922:s[6-7] = p_linear_phi %914:s[6-7], %920:s[6-7]
	p_logical_start
	p_logical_end
	 s2: %471:s[14-15],   b: %472:scc,  s2: %473:exec = s_and_saveexec_b64 %922:s[6-7]
	 s2: %474:s[6-7],   b: %475:scc = s_andn2_b64 %471:s[14-15], %922:s[6-7]
	p_cbranch_z %473:exec BB13, BB12
BB12
/* logical preds: BB11, / linear preds: BB11,  */
	p_logical_start
	p_logical_end
	p_branch BB14
BB13
/* logical preds: / linear preds: BB11,  */
	p_branch BB14
BB14
/* logical preds: / linear preds: BB12, BB13,  */
	 s2: %476:s[14-15] = s_mov_b64 %0:exec
	 s2: %927:exec = p_parallelcopy %474:s[6-7]
	p_cbranch_z %927:exec BB16, BB15
```

**Comment by daniel-schuermann** [comment]
oh, did it hit the assert(branch->opcode == aco_opcode::p_branch); ? (see BB11 has p_cbranch_z)
might come from the change to do this backwards, but I don't understand we don't hit this assert on any other game. it looks like the normal pattern.

**Comment by tarceri** [comment]
No it is trying to merge BB11 with BB14 as BB12 and BB13 are empty. But we hit:

```
      assert(instr->opcode == aco_opcode::p_linear_phi ||
             instr->opcode == aco_opcode::s_mov_b64 ||
             instr->opcode == aco_opcode::p_parallelcopy);
```
Update: oh I see there are other asserts in this function. I should have been more specific when reporting the bug.

**Comment by tarceri** [comment]
To be specific we hit the assert when we see: s_and_saveexec_b64

Update: oh actually I think the p_logical_start and p_logical_end cause it too (I just added them to the assert when I was testing add one point).

**Comment by daniel-schuermann** [comment]
ah, I see. So, when it hits the assert, BB12, 13, 14 have been removed already. So, it's indeed a different pattern. The code thinks, it's a merge block, but it actually is the branch block. ok, thx!

**Comment by daniel-schuermann** [comment]
Fixed by 5850a9dc3a4d1078703325902f710e947373195f with TODO to optimize.
